packages/playwright-core/src/server/injected/injectedScript.ts

```diff
@@ -1099,7 +1099,10 @@ export class InjectedScript {
       // Single text value.
       let received: string | undefined;
       if (expression === 'to.have.attribute') {
-        received = element.getAttribute(options.expressionArg) || '';
+        const value = element.getAttribute(options.expressionArg);
```

Comment:
I would drop '' and leave the rest as is. That way users get 'got null' in the error message

Reply:
Received cannot be null, so we'd have to covert it to undefined in that case which would result in 'Unknown expect matcher: to.have.attribute' which is not what we want to see. Additionally it should return true for `.not.toHaveAttribute('foo', '');` rather than throw when the attribute is missing.

Reply:
> Received cannot be null

Why?

Reply:
Updated